core/src/main/scala/cromwell/core/retry/IORetry.scala

```diff
@@ -0,0 +1,52 @@
+package cromwell.core.retry
+
+import cats.effect.{IO, Timer}
+
+import scala.concurrent.duration._
+
+object IORetry {
+  def noOpOnRetry[S]: (Throwable, S) => S = (_, s) => s
+
+  /**
+    * When we reach a point where we need to fail the IO (because we ran out of retries, or exception was fatal etc...)
+    * This provides a method to generate a Throwable from the current state and the last failure.
+    */
+  trait StatefulIoError[S] {
+    def toThrowable(state: S, throwable: Throwable): Throwable
+  }
+
+  /**
+    * Same as Retry.withRetry using cats.effect.IO
+    * The state is being passed down upon each retry. It can be use to accumulate information as IO operations
+    * fail and are being retried.
+    */
+  def withRetry[A, S](io: IO[A],
+                      state: S,
+                      maxRetries: Option[Int],
+                      backoff: Backoff = SimpleExponentialBackoff(5.seconds, 10.seconds, 1.1D),
```

Comment:
TIL: [`backoff` is actually one word not two](https://en.wikipedia.org/wiki/Exponential_backoff) and is [misspelled in google's libs](https://github.com/google/google-http-java-client/blob/1.23.0/google-http-client/src/main/java/com/google/api/client/util/BackOffUtils.java).